tests/integration/metrics_tests/test_metrics.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+import time
+import yaml
+
+import ops
+import pytest
+import urllib3
+from pytest_operator.plugin import OpsTest
+from pathlib import Path
+from ..ha_tests import helpers as ha_helpers
+
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+DATABASE_APP_NAME = "mongodb-k8s"
+MONGODB_EXPORTER_PORT = 9216
+MEDIAN_REELECTION_TIME = 12
+
+
+@pytest.fixture(scope="module")
+def chaos_mesh(ops_test: OpsTest) -> None:
+    ha_helpers.deploy_chaos_mesh(ops_test.model.info.name)
+
+    yield
+
+    ha_helpers.destroy_chaos_mesh(ops_test.model.info.name)
+
+
+async def get_address(ops_test: OpsTest, app_name=DATABASE_APP_NAME, unit_num=0) -> str:
+    """Get the address for a unit."""
+    status = await ops_test.model.get_status()  # noqa: F821
+    address = status["applications"][app_name]["units"][f"{app_name}/{unit_num}"]["address"]
+    return address
+
+
+async def verify_endpoints(ops_test: OpsTest, unit: ops.model.Unit) -> str:
+    """Verifies mongodb endpoint is functional on a given unit."""
+    http = urllib3.PoolManager()
+
+    unit_address = await get_address(ops_test=ops_test)
+    mongodb_exporter_url = f"http://{unit_address}:{MONGODB_EXPORTER_PORT}/metrics"
+    mongo_resp = http.request("GET", mongodb_exporter_url)
+
+    assert mongo_resp.status == 200
+
+    # if configured correctly there should be more than one mongodb metric present
+    mongodb_metrics = mongo_resp._body.decode("utf8")
+    assert mongodb_metrics.count("mongo") > 10
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy three units of MongoDB and one unit of TLS."""
+    async with ops_test.fast_forward():
+        my_charm = await ops_test.build_charm(".")
+        resources = {"mongodb-image": METADATA["resources"]["mongodb-image"]["upstream-source"]}
+        await ops_test.model.deploy(my_charm, num_units=3, resources=resources, series="jammy")
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=2000)
+
+
+async def test_endpoints(ops_test: OpsTest):
+    """Sanity check that endpoints are running."""
+    application = ops_test.model.applications[DATABASE_APP_NAME]
+
+    for unit in application.units:
+        await verify_endpoints(ops_test, unit)
+
+
+async def test_endpoints_network_cut(ops_test: OpsTest, chaos_mesh):
+    """Verify that endpoint still function correctly after a network cut."""
+    # retrieve a primary unit and a non-primary unit (active-unit). The primary unit will have its
+    # network disrupted, while the active unit allows us to communicate to `mongod`
+    primary = await ha_helpers.get_replica_set_primary(ops_test)
+    active_unit = [
+        unit
+        for unit in ops_test.model.applications[DATABASE_APP_NAME].units
+        if unit.name != primary.name
```

Comment:
maybe we can use `filter` here?

Reply:
The generator (current) style has good readability (but inconsistent performance).
I think it is ok to keep it as is.